Add list_folders method for gcs

razvanvilceanu opened the pull request:
<!-- Thanks for contributing 🎉! Please ensure the title neatly summarizes the proposed changes. -->

<!-- Overview -->

Closes https://github.com/PrefectHQ/prefect-gcp/issues/118

### Example
<!-- A code blurb is best. Changes to features should include an example that is executable by a new user. -->

### Screenshots
<!--
Any relevant screenshots
  - The updated docs page from `mkdocs serve`.
  - Output from running the example.
  - Service integration test results.
-->

### Checklist
<!-- These boxes may be checked after opening the pull request. -->

- [ ] References any related issue by including "Closes #<Issue Number>" or "Closes <Issue URL>".
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect-gcp/issues/new/choose) first.
- [ ] Includes tests or only affects documentation.
- [ ] Passes `pre-commit` checks.
  - Run `pre-commit install && pre-commit run --all` locally for formatting and linting.
- [ ] Includes screenshots of documentation updates.
  - Run `mkdocs serve` view documentation locally.
- [ ] Summarizes PR's changes in [CHANGELOG.md](https://github.com/PrefectHQ/prefect-gcp/blob/main/CHANGELOG.md)

ahuang11 commented on CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- list_folders method for GcsBucket - [#issue](https://github.com/PrefectHQ/prefect-gcp/issues/118)
```
```suggestion
- list_folders method for GcsBucket - [#121](https://github.com/PrefectHQ/prefect-gcp/issues/121)

```

ahuang11 commented on prefect_gcp/cloud_storage.py:
```diff
@@ -753,6 +753,48 @@ async def list_blobs(self, folder: str = "") -> List["Blob"]:
+            client.list_blobs, self.bucket, prefix=bucket_path
+        )
+        blobs = list(blobs)
+        if bucket_path is not None:
+            unique_folders = list(set([blob.name.split(os.sep)[1] for blob in blobs if "." not in blob.name.split(os.sep)[1]]))
+            return [folder for folder in unique_folders if folder]
+        else:
+            return list(set([blob.name.split(os.sep)[0] for blob in blobs if blob.name.endswith(os.sep)]))
```
I am having a hard time understanding this logic. Can you elaborate on the intent?

Is "." checking to see if it's a file?

Also, I don't think we should use os.sep since Windows will use "\" which isn't used in GCP.

razvanvilceanu replied:
Yes, the set(list()) is for taking only the unique elements because there were folders showing up for each blob (e.g. ['subdirectory/file1.txt', 'subdirectory/file2.txt' ]  and the output would be subdirectory*2)

The '.' is indeed for files because these we're showing up if they were in the same directory as a subdirectory.

And the split()[0] and split()[1] is for showing the root directory or the 2nd level one

razvanvilceanu replied:
We can replace os.sep with '/' as I didn't know that about Windows

ahuang11 replied:
We can probably simplify by:
1. using a set comprehension directly
2. using PurePosixPath to handle discovery of folders
3. removing the if/else

Something like this:
```
from prefect_gcp.cloud_storage import GcsBucket
from pathlib import PurePosixPath

bucket = await GcsBucket.load("my-bucket")
blobs = client.list_blobs("my-sbx-bucket")
{str(PurePosixPath(blob.name).parent) for blob in blobs}
```

razvanvilceanu commented:
These suggestions sound pretty great! I haven't used PurePosixPath so far, but I'll handle it. However, wouldn't be better to keep the final return state as a list to benefit from existing methods like sort? I think of something like list({set_comptehension}).

I'll come back tomorrow with updates! Thank you!

________________________________
From: Andrew ***@***.***>
Sent: Wednesday, January 25, 2023 8:34:41 PM
To: PrefectHQ/prefect-gcp ***@***.***>
Cc: razvanvilceanu ***@***.***>; Author ***@***.***>
Subject: Re: [PrefectHQ/prefect-gcp] Add listidr method for gcs (PR #121)


@ahuang11 commented on this pull request.

________________________________

In prefect_gcp/cloud_storage.py<https://github.com/PrefectHQ/prefect-gcp/pull/121#discussion_r1087030845>:

> +        if bucket_path is not None:
+            unique_folders = list(set([blob.name.split(os.sep)[1] for blob in blobs if "." not in blob.name.split(os.sep)[1]]))
+            return [folder for folder in unique_folders if folder]
+        else:
+            return list(set([blob.name.split(os.sep)[0] for blob in blobs if blob.name.endswith(os.sep)]))


We can probably simplify by:

  1.  using a set comprehension directly
  2.  using PurePosixPath to handle discovery of folders
  3.  removing the if/else

Something like this:

from prefect_gcp.cloud_storage import GcsBucket
from pathlib import PurePosixPath

bucket = await GcsBucket.load("my-bucket")
blobs = client.list_blobs("my-sbx-bucket")
{str(PurePosixPath(blob.name).parent) for blob in blobs}


—
Reply to this email directly, view it on GitHub<https://github.com/PrefectHQ/prefect-gcp/pull/121#discussion_r1087030845>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AJSHYZDWGMQSANRBZZ76BGLWUFW4DANCNFSM6AAAAAAUGEULDM>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

ahuang11 commented:
> However, wouldn't be better to keep the final return state as a list to benefit from existing methods like sort? I think of something like list({set_comptehension}). 

Sure, at the end, you can do `sorted({...})` to sort and convert to list.

razvanvilceanu commented:
Fixed the tests and ran pre-commit and everything passes, is there anything else that runs on git workflow that I should take care of?

ahuang11 commented on tests/test_cloud_storage.py:
```diff
@@ -210,6 +210,17 @@ def gcs_bucket_with_bucket_folder(self, gcp_credentials, request):
+    def test_list_folders_with_root_only(self, gcs_bucket_with_bucket_folder):
+        blobs = gcs_bucket_with_bucket_folder.list_folders()
+        assert len(blobs) == 2
+        assert blobs[0] == "base_folder"
+        assert blobs[1] == "base_folder/sub_folder"
```
```suggestion
        assert set(blobs) == {"base_folder", "base_folder/sub_folder"}
```

ahuang11 commented:
I applied a quick fix to the tests; there's nothing else to take care of in this PR now. Thank you for contributing!

razvanvilceanu commented:
thanks for the help @ahuang11, couldn't have done it without you
